dataflow/src/ops/project.rs

```diff
@@ -122,6 +124,79 @@ impl Ingredient for Project {
         vec![self.src.as_global()]
     }
 
+    fn can_query_through(&self) -> bool {
+        true
+    }
+
+    fn query_through<'a>(
+        &self,
+        columns: &[usize],
+        key: &KeyType,
+        states: &'a StateMap,
+    ) -> Option<Option<Box<Iterator<Item = Cow<'a, [DataType]>> + 'a>>> {
+        let emit = self.emit.clone();
+        let additional = self.additional.clone();
+        let expressions = self.expressions.clone();
+        states
+            .get(&*self.src)
+            .and_then(move |state| match state.lookup(columns, key) {
+                LookupResult::Some(Cow::Borrowed(rs)) => {
+                    let r = match emit {
+                        Some(emit) => Box::new(rs.iter().map(move |r| {
+                            let mut new_r = Vec::with_capacity(r.len());
+                            for i in emit.iter() {
+                                new_r.push(r[*i].clone());
+                            }
+
+                            if let Some(ref e) = expressions {
+                                new_r.extend(e.into_iter().map(|i| eval_expression(i, &r[..])));
+                            }
+
+                            if let Some(ref a) = additional {
+                                new_r.append(&mut a.clone());
+                            }
+
+                            Cow::from(new_r)
+                        })) as Box<_>,
+                        None => Box::new(rs.iter().map(|r| Cow::from(&r[..]))) as Box<_>,
+                    };
+
+                    Some(Some(r))
+                }
+                LookupResult::Some(Cow::Owned(rs)) => {
+                    let r = match emit {
+                        Some(emit) => Box::new(rs.into_iter().map(move |r| {
+                            let r = r.unpack();
+                            let mut new_r = Vec::with_capacity(r.len());
+                            let mut expr: Vec<DataType> = if let Some(ref e) = expressions {
+                                e.into_iter().map(|i| eval_expression(i, &r[..])).collect()
+                            } else {
+                                vec![]
+                            };
+
+                            new_r.extend(
+                                r.into_iter()
+                                    .enumerate()
+                                    .filter(|(i, _)| emit.iter().any(|e| e == i))
+                                    .map(|(_, c)| c),
+                            );
+
+                            new_r.append(&mut expr);
+                            if let Some(ref a) = additional {
+                                new_r.append(&mut a.clone());
+                            }
+
+                            Cow::from(new_r)
+                        })) as Box<_>,
+                        None => Box::new(rs.into_iter().map(|r| Cow::from(r.unpack()))) as Box<_>,
+                    };
+
+                    Some(Some(r))
+                }
```

Comment:
The repetition here is pretty sad :/ Will this get better with the `into_iter` from #76?